pyiceberg/table/__init__.py

```diff
@@ -533,6 +535,8 @@ def update_table_metadata(base_metadata: TableMetadata, updates: Tuple[TableUpda
     for update in updates:
         new_metadata = _apply_table_update(update, new_metadata, context)
 
+    # Rebuild metadata to trigger validation
+    new_metadata = TableMetadataUtil.parse_obj(copy(new_metadata.model_dump()))
```

Comment:
Since `model_copy` performs a shallow copy by default, I believe we need to execute a deep copy before returning the final `new_metadata`. Otherwise, `base_metadata` might be inadvertently altered due to any improper updates applied to `new_metadata` subsequently.

Furthermore, as indicated by https://github.com/pydantic/pydantic/issues/418 and by some local tests, `model_copy(update=)` does not validate the contents of update. I think it might be good to reconstruct the metadata at this point to initiate the validation process.

(Alternatively, we could perform a deep copy here and incorporate the validation into our unit tests. Open to discussion on this approach.)

Reply:
I think we're okay to skip the validation on the Pydantic side, we should validate the input anyway to generate a more meaningful error.

We could set it to deep-copy using the `deep=True` kwarg. Doing this in Pydantic is probably an order of magnitude more efficient than doing this in Python.

Reply:
Thanks for the explanation! I changed to `model_copy(deep=True)` here and moved the validation to a unit test